.. _cfdnapro_usage_tutorials:

********************************************************
cfDNAPro General Usage
********************************************************

.. contents:: Table of Contents

--------------------------------------------------------------------------

Insert Size Metrics Data
=======================================

In this tutorial, we will be working with fragment size metrics data, which
is commonly generated in cfDNA research using
`Picard <https://samtools.github.io/hts-specs/SAMv1.pdf>`__.


    Picard is a set of command line tools for manipulating high-throughput sequencing (HTS) data and formats such as SAM/BAM/CRAM and VCF. These file formats are defined in the Hts-specs repository. See especially the SAM specification and the VCF specification. [1]_

Fragment size, also commonly referred to as cfDNA fragment length,
can be directly extracted from BAM files using ``cfDNAPro``,
or alternatively, the package can read insert size metric files
generated by the Picard tool.
This tutorial will specifically focus on
processing insert size metrics from .txt files,
which can be generated using the `CollectInsertSizeMetrics <https://broadinstitute.github.io/picard/command-line-overview.html#CollectInsertSizeMetrics>`__ function.


Processing .txt Files
=========================
To utilize the cfDNAPro package, organize all .txt files generated by Picard 
into sub-folders named according to each cohort even if there is only one cohort.
Example .txt files are included with the installation of this package.

Below are the steps to access the example data:

.. code:: R

    library(cfDNAPro)
    # Get the path to example data in the cfDNAPro package library path.
    data_path <- examplePath("groups_picard")

The example data is organized into four groups,
each stored in its own sub-folder named
``cohort_1``, ``cohort_2``, ``cohort_3``, and ``cohort_4``.
Each folder contains specific samples,
represented as .txt files.

To review the files within these directories,
use the list.files function with the following parameters:
``list.files(data_path, full.names = TRUE, recursive = TRUE)``

Note: The cohorts labeled as ``cohort_1``, ``cohort_2``, ``cohort_3``,
and ``cohort_4`` are not associated with any biological significance.
These are simulated datasets not linked to any specific cancer types.
In the following sections, you will encounter several plots
demonstrating biological cfDNA feature patterns.
Please note that these plots and data are intended solely to illustrate
the functionality of cfDNAPro for data analysis and visualization
purposes.

In your own analysis,
you can apply these functions to analyze cfDNA NGS sequencing data
from cancer samples, helping you uncover fragmentation features that
are relevant to your research questions.

.. [1] Source: Official Picard website. URL: http://broadinstitute.github.io/picard/
